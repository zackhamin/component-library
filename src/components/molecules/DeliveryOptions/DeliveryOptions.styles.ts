import { colors } from 'Styles';
import styled from 'styled-components';

const StyledList = styled.ul`
	list-style: none;
	display: flex;
	flex-direction: column;
	gap: 1.6rem;
	padding: 0;
	margin: 0;

	.free-option {
		margin-left: 1.6rem;
		color: ${colors.pomegranate};
	}
`;

export { StyledList };
