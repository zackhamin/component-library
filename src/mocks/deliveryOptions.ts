import { Delivery } from '@Types/delivery';

export const mockDeliveryOptions: Delivery[] = [
	{
		id: 'NEXT_DAY_DELIVERY',
		optionId: '98a820f3-3f0c-431d-925a-b1bebb1c0bfd',
		name: 'Next Day Delivery',
		price: {
			amount: '4.99',
			currency: 'GBP',
		},
		cutoff: '2022-11-01T21:00:00.000Z',
		timeWindow: {
			from: '2022-11-02T08:00:00.000Z',
			to: '2022-11-02T21:00:00.000Z',
		},
	},
	{
		id: 'STANDARD_DELIVERY',
		optionId: 'fdd1b6db-1504-4667-b125-d4e59f62bb8e',
		name: 'Standard Delivery',
		price: {
			amount: '3.49',
			currency: 'GBP',
		},
		cutoff: '2022-11-01T20:00:00.000Z',
		timeWindow: {
			from: '2022-11-04T08:00:00.000Z',
			to: '2022-11-05T21:00:00.000Z',
		},
	},
	{
		id: 'NOMINATED_DAY',
		optionId: 'dad5be1f-6c5a-4f99-b768-9af0bdab566b',
		name: 'Nominated Day Delivery',
		price: {
			amount: '4.99',
			currency: 'GBP',
		},
		cutoff: '2022-11-02T20:00:00.000Z',
		timeWindow: {
			from: '2022-11-03T08:00:00.000Z',
			to: '2022-11-03T23:59:59.999Z',
		},
	},
	{
		id: 'NOMINATED_DAY',
		optionId: 'cc801e2d-bbfb-4289-8b8f-9d33959fd47e',
		name: 'Nominated Day Delivery',
		price: {
			amount: '4.99',
			currency: 'GBP',
		},
		cutoff: '2022-11-03T20:00:00.000Z',
		timeWindow: {
			from: '2022-11-04T08:00:00.000Z',
			to: '2022-11-04T23:59:59.999Z',
		},
	},
];
