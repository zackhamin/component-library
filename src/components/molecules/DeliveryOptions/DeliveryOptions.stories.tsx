import { mockDeliveryOptions } from 'MockData/deliveryOptions';
import { DeliveryOptions, DeliveryOptionsProps } from './DeliveryOptions';

export default {
	title: 'Molecules/DeliveryOptions',
	component: DeliveryOptions,
};

export const DeliveryOptionsStory = (args: DeliveryOptionsProps) => (
	<DeliveryOptions {...args} />
);

DeliveryOptionsStory.storyName = 'DeliveryOptions';
DeliveryOptionsStory.args = {
	deliveryOptions: mockDeliveryOptions,
	selectedDeliveryOption: mockDeliveryOptions[0],
};
DeliveryOptionsStory.argTypes = {
	handleSelect: { action: 'selected' },
};
