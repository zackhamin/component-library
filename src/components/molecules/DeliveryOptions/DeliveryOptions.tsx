import { Delivery } from '@Types/delivery';
import { StyledList } from './DeliveryOptions.styles';

import { Heading } from 'Atoms/Heading';
import { ListOption } from 'Atoms/ListOption';
import { Price } from 'Atoms/Price';
import { Text } from 'Atoms/Text';
import { dateToDayWithTime, dateToWeekdayWithMonth } from 'Utils/date';

export interface DeliveryOptionsProps {
	deliveryOptions: Delivery[];
	selectedDeliveryOption?: Delivery;
	handleSelect: (option: Delivery) => void;
	i18n?: {
		locale: string;
	};
}

export const DeliveryOptions = ({
	deliveryOptions,
	selectedDeliveryOption,
	handleSelect,
	i18n,
}: DeliveryOptionsProps) => {
	return (
		<StyledList>
			{deliveryOptions.map((option) => {
				const isSelected = option.optionId === selectedDeliveryOption?.optionId;

				return (
					<ListOption
						key={option.optionId}
						isSelected={isSelected}
						handleSelect={() => handleSelect(option)}
					>
						<Heading size="1.8rem">{option.name}</Heading>
						<Text>
							Order before{' '}
							{dateToDayWithTime(new Date(option.cutoff), i18n?.locale, true)}
						</Text>
						<Text>
							Delivery on or before{' '}
							{dateToWeekdayWithMonth(new Date(option.timeWindow.to))}
						</Text>
						<Text>
							<Price {...option.price} />
							{option.freeOption && (
								<span className="free-option">
									Free from <Price {...option.freeOption} />
								</span>
							)}
						</Text>
					</ListOption>
				);
			})}
		</StyledList>
	);
};
