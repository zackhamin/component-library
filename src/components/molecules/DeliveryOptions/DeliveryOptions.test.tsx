import { fireEvent, render } from '@testing-library/react';

import { DeliveryOptions } from './DeliveryOptions';
import { mockDeliveryOptions } from 'MockData/deliveryOptions';

jest.spyOn(console, 'error').mockImplementation(jest.fn);

describe('DeliveryOptions', () => {
	const mockSelectHandler = jest.fn();

	afterEach(() => {
		jest.resetAllMocks();
	});

	it('renders options correctly', () => {
		const { container } = render(
			<DeliveryOptions
				handleSelect={mockSelectHandler}
				deliveryOptions={mockDeliveryOptions}
				selectedDeliveryOption={mockDeliveryOptions[0]}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('updates selected address option', () => {
		const { container } = render(
			<DeliveryOptions
				handleSelect={mockSelectHandler}
				deliveryOptions={mockDeliveryOptions}
			/>
		);

		const optionToSelect = container.getElementsByTagName('li')[1];
		expect(optionToSelect).not.toHaveClass('active');
		fireEvent.click(optionToSelect);
		setTimeout(() => {
			expect(optionToSelect).toHaveClass('active');
		});
	});

	it('calls the select handler correctly', () => {
		const { container } = render(
			<DeliveryOptions
				handleSelect={mockSelectHandler}
				deliveryOptions={mockDeliveryOptions}
				selectedDeliveryOption={mockDeliveryOptions[0]}
			/>
		);

		const methodToSelect = container.getElementsByTagName('li')[1];

		fireEvent.click(methodToSelect);

		expect(mockSelectHandler).toHaveBeenCalled();
	});
});
