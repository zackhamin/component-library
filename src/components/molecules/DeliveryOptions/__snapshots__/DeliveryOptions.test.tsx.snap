// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeliveryOptions renders options correctly 1`] = `
.c0 {
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1.6rem;
  padding: 0;
  margin: 0;
}

.c0 .free-option {
  margin-left: 1.6rem;
  color: #F7403A;
}

.c2 {
  margin: 0;
  font-family: BrandonGrotesque-Bold,Arial,Helvetica,sans-serif;
  color: #333333;
  font-size: 1.8rem;
  line-height: 140%;
  font-weight: 450;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  list-style: none;
  border: 1px solid #EEEEEF;
  background: #ffffff;
  border-radius: 8px;
  padding: 1.6rem;
  cursor: pointer;
}

.c1.active {
  border: 1px solid #7FC1BA;
  background: rgba(3,134,120,0.1);
}

.c1.active .radio {
  border: 1px solid #00594C;
}

.c1.active .radio > span {
  background: #00594C;
}

.c1 .details {
  margin-left: 2.6rem;
}

.c1 .radio {
  display: block;
  border: 1px solid #bebebe;
  border-radius: 50%;
}

.c1 .radio > span {
  display: block;
  width: 10px;
  height: 10px;
  margin: 3px;
  border-radius: 50%;
  background: transparent;
}

.c3 {
  margin: 0;
  font-weight: 400;
  font-family: Helvetica,Arial,sans-serif;
  color: #333333;
  font-size: 1.6rem;
  line-height: 160%;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1 active"
    >
      <div
        class="radio"
      >
        <span />
      </div>
      <div
        class="details"
      >
        <h2
          class="c2"
        >
          Next Day Delivery
        </h2>
        <p
          class="c3"
        >
          Order before
           
          9 pm
        </p>
        <p
          class="c3"
        >
          Delivery on or before
           
          Wed, 2 Nov
        </p>
        <p
          class="c3"
        >
          £4.99
        </p>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="radio"
      >
        <span />
      </div>
      <div
        class="details"
      >
        <h2
          class="c2"
        >
          Standard Delivery
        </h2>
        <p
          class="c3"
        >
          Order before
           
          8 pm
        </p>
        <p
          class="c3"
        >
          Delivery on or before
           
          Sat, 5 Nov
        </p>
        <p
          class="c3"
        >
          £3.49
        </p>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="radio"
      >
        <span />
      </div>
      <div
        class="details"
      >
        <h2
          class="c2"
        >
          Nominated Day Delivery
        </h2>
        <p
          class="c3"
        >
          Order before
           
          8 pm
        </p>
        <p
          class="c3"
        >
          Delivery on or before
           
          Thu, 3 Nov
        </p>
        <p
          class="c3"
        >
          £4.99
        </p>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="radio"
      >
        <span />
      </div>
      <div
        class="details"
      >
        <h2
          class="c2"
        >
          Nominated Day Delivery
        </h2>
        <p
          class="c3"
        >
          Order before
           
          8 pm
        </p>
        <p
          class="c3"
        >
          Delivery on or before
           
          Fri, 4 Nov
        </p>
        <p
          class="c3"
        >
          £4.99
        </p>
      </div>
    </li>
  </ul>
</div>
`;
